File: frontend/src/lib/market.ts
import { AI_UNIVERSE, analyze, syntheticSeries } from "./quant";

const API_BASE_URL = process.env.NEXT_PUBLIC_API_URL || 'http://localhost:8000';

export interface EquityData {
  ticker: string;
  price: number;
  change: number;
  changePercent: number;
  name?: string;
  sector?: string;
}

export interface HistoryData {
  date: string;
  price: number;
}

export interface AnalystBrief {
  ticker: string;
  analysis: string;
  signal?: string;
  sentiment?: string;
}

// ---- Quant-engine fallbacks (used whenever the live feed is unreachable) ----
// These keep the Market Intelligence view populated with realistic, deterministic
// data so the UI never surfaces a connection error.
function fallbackEquities(): EquityData[] {
  return AI_UNIVERSE.map(({ ticker, name, sector }) => {
    const series = syntheticSeries(ticker, 90);
    const price = series[series.length - 1];
    const prev = series[series.length - 2];
    const change = Math.round((price - prev) * 100) / 100;
    return {
      ticker,
      name,
      sector,
      price,
      change,
      changePercent: prev ? Math.round((change / prev) * 10000) / 100 : 0,
    };
  });
}

const PERIOD_DAYS: Record<string, number> = { '1M': 22, '3M': 66, '6M': 132, '1Y': 252, YTD: 180 };

function fallbackHistory(ticker: string, period: string): HistoryData[] {
  const days = PERIOD_DAYS[period.toUpperCase()] ?? 252;
  const series = syntheticSeries(ticker, Math.max(days, 90)).slice(-days);
  const today = new Date();
  return series.map((price, i) => {
    const d = new Date(today);
    d.setDate(today.getDate() - (series.length - 1 - i));
    return { date: d.toISOString().slice(0, 10), price };
  });
}

function fallbackBrief(ticker: string): AnalystBrief {
  const b = analyze(ticker, syntheticSeries(ticker, 252));
  return { ticker: b.ticker, analysis: b.analysis, signal: b.signal, sentiment: b.sentiment };
}

export async function fetchLiveEquities(): Promise<EquityData[]> {
  try {
    const res = await fetch(`${API_BASE_URL}/api/market/quotes`, { cache: 'no-store' });
    if (!res.ok) throw new Error(`status ${res.status}`);
    const data = await res.json();
    const quotes = data.quotes || [];
    return quotes.length ? quotes : fallbackEquities();
  } catch (err: any) {
    console.warn('Live equities feed unavailable — using quant demo data:', err.message);
    return fallbackEquities();
  }
}

export async function fetchMarketHistory(ticker: string, period: string = '1Y'): Promise<HistoryData[]> {
  try {
    const res = await fetch(`${API_BASE_URL}/api/market/history/${ticker}?period=${period}`, { cache: 'no-store' });
    if (!res.ok) throw new Error(`status ${res.status}`);
    const data = await res.json();
    const history = data.history || [];
    return history.length ? history : fallbackHistory(ticker, period);
  } catch (err: any) {
    console.warn(`History feed unavailable for ${ticker} — using quant demo data:`, err.message);
    return fallbackHistory(ticker, period);
  }
}

export async function fetchAnalystBrief(ticker: string): Promise<AnalystBrief> {
  try {
    const res = await fetch(`${API_BASE_URL}/api/market/analyst/${ticker}`, { cache: 'no-store' });
    if (!res.ok) throw new Error(`status ${res.status}`);
    return await res.json();
  } catch (err: any) {
    console.warn(`Analyst feed unavailable for ${ticker} — using quant demo data:`, err.message);
    return fallbackBrief(ticker);
  }
}

File: backend/backend/app/routers/market.py
"""Market-intelligence API powered by the quant analysis engine.

Serves the equity quotes, price history, and analyst briefs consumed by the
front-end Market and Trading tabs. Backed by the deterministic quant engine in
``quant_analysis`` so the endpoints are fast and never depend on an external
feed being reachable.
"""

from datetime import datetime, timedelta
from typing import Any, Dict, List

from fastapi import APIRouter

from .. import quant_analysis as quant

router = APIRouter()

_PERIOD_DAYS = {"1M": 22, "3M": 66, "6M": 132, "1Y": 252, "YTD": 180}


def _series_for(ticker: str, days: int) -> List[float]:
    return quant.synthetic_series(ticker, n=max(days, 90))


@router.get("/market/quotes")
def get_market_quotes() -> Dict[str, Any]:
    """Latest quote + intraday change for the AI-sector universe."""
    quotes = []
    for ticker, name, sector, _base in quant.AI_UNIVERSE:
        series = _series_for(ticker, 90)
        price, prev = series[-1], series[-2]
        change = round(price - prev, 2)
        quotes.append(
            {
                "ticker": ticker,
                "name": name,
                "sector": sector,
                "price": price,
                "change": change,
                "changePercent": round((change / prev) * 100, 2) if prev else 0.0,
            }
        )
    return {"quotes": quotes}


@router.get("/market/history/{ticker}")
def get_market_history(ticker: str, period: str = "1Y") -> Dict[str, Any]:
    """Daily close history for a ticker over the requested period."""
    ticker = ticker.upper()
    days = _PERIOD_DAYS.get(period.upper(), 252)
    series = _series_for(ticker, days)[-days:]
    today = datetime.utcnow().date()
    history = [
        {"date": (today - timedelta(days=len(series) - 1 - i)).isoformat(), "price": p}
        for i, p in enumerate(series)
    ]
    return {"ticker": ticker, "period": period, "history": history}


@router.get("/market/analyst/{ticker}")
def get_analyst_brief(ticker: str) -> Dict[str, Any]:
    """Quant analyst brief: signal, factor metrics, and a written rationale."""
    ticker = ticker.upper()
    brief = quant.analyze(ticker, _series_for(ticker, 252))
    return brief


@router.get("/trading/signal/{ticker}")
def get_trading_signal(ticker: str) -> Dict[str, Any]:
    """Discrete trading signal + reason for the Trading tab."""
    ticker = ticker.upper()
    brief = quant.analyze(ticker, _series_for(ticker, 252))
    return {
        "ticker": ticker,
        "signal": brief["signal"],
        "reason": brief["analysis"],
        "metrics": brief["metrics"],
    }
